src/storage/buffer_manager/buffer_manager.cpp

```diff
@@ -33,61 +49,215 @@ void BufferManager::resize(uint64_t newSizeForDefaultPagePool, uint64_t newSizeF
+    }
 }
 
 void BufferManager::unpin(BufferManagedFileHandle& fileHandle, page_idx_t pageIdx) {
-    return fileHandle.isLargePaged() ? bufferPoolLargePages->unpin(fileHandle, pageIdx) :
-                                       bufferPoolDefaultPages->unpin(fileHandle, pageIdx);
+    fileHandle.acquirePageLock(pageIdx, LockMode::SPIN);
+    unpinWithoutAcquiringPageLock(fileHandle, pageIdx);
+    fileHandle.releasePageLock(pageIdx);
+}
+
+void BufferManager::unpinWithoutAcquiringPageLock(
+    BufferManagedFileHandle& fileHandle, common::page_idx_t pageIdx) {
+    auto pageState = fileHandle.getPageState(pageIdx);
+    // `count` is the value of `pinCount` before sub.
+    auto count = pageState->decrementPinCount();
+    assert(count >= 1);
+    if (count == 1) {
+        addToEvictionQueue(pageState);
+    }
+}
+
+void BufferManager::resize(uint64_t newSize) {
+    if (newSize < usedMemory) {
+        throw BufferManagerException("Resizing to a smaller Buffer Pool Size is unsupported.");
+    }
+    maxMemory = newSize;
+}
+
+// This function tries to load the given page into a frame. Due to our design of mmap, each page is
+// uniquely mapped to a virtual memory buffer (frame). Thus, claiming a frame is equivalent to
+// ensuring enough physical memory is available.
+// First, we reserve the memory for the page, which increments the atomic counter `usedMemory`.
+// Then, we check if there is enough memory available. If not, we evict pages until we have enough
+// or we can find no more pages to be evicted.
+// Lastly, we double check if the needed memory is available. If not, we free the memory we reserved
+// and return false, otherwise, we load the page to its corresponding frame and return true.
+bool BufferManager::claimAFrame(BufferManagedFileHandle& fileHandle, common::page_idx_t pageIdx,
+    PageReadPolicy pageReadPolicy) {
+    page_offset_t pageSizeToClaim = fileHandle.getPageSize();
+    // Reserve the memory for the page.
+    auto currentUsedMem = reserveUsedMemory(pageSizeToClaim);
+    EvictionCandidate evictionCandidate;
+    uint64_t claimedMemory = 0;
+    // Evict pages if necessary until we have enough memory.
+    while ((currentUsedMem + pageSizeToClaim - claimedMemory) > maxMemory.load()) {
+        if (!evictionQueue->dequeue(evictionCandidate)) {
+            // Cannot find more pages to be evicted. Free the memory we reserved and return false.
+            freeUsedMemory(pageSizeToClaim);
+            return false;
+        }
+        if (evictionCandidate.evictionTimestamp !=
+                evictionCandidate.pageState->getEvictionTimestamp() ||
+            evictionCandidate.pageState->getPinCount() > 0) {
+            continue;
+        }
+        // Now we find a block whose pin count can be 0, and potentially haven't been accessed since
+        // enqueued. Try to lock the page and double check if the page's pin count is actually 0 and
```

Comment:
Try to lock -> We try to evict the page from its vm page by calling tryEvictPage(). which will check that the page's pin count is .., if so, we evict ...

Reply:
Done.